collect_app/src/androidTest/java/org/odk/collect/android/support/ResetStateRule.java

```diff
@@ -29,14 +56,44 @@ public Statement apply(final Statement base, Description description) {
 
         @Override
         public void evaluate() throws Throwable {
+            // Reset any singleton state
+            if (appDependencyModule != null) {
+                CollectHelpers.overrideAppDependencyModule(appDependencyModule);
+            } else {
+                CollectHelpers.overrideAppDependencyModule(new AppDependencyModule());
+            }
+
+            // Make sure we clear all our shared prefs - ignore logic that doesn't reset keys
+            Context context = InstrumentationRegistry.getInstrumentation().getTargetContext();
+            PreferenceManager.getDefaultSharedPreferences(context).edit().clear().commit();
+            context.getSharedPreferences(ADMIN_PREFERENCES, 0).edit().clear().commit();
+            new MetaSharedPreferencesProvider(context).getMetaSharedPreferences().edit().clear().commit();
+
+            // Reset the app in both the old and new storage locations (just nuke dirs)
             List<Integer> resetActions = Arrays.asList(
-                    ResetUtility.ResetAction.RESET_PREFERENCES, ResetUtility.ResetAction.RESET_INSTANCES,
-                    ResetUtility.ResetAction.RESET_FORMS, ResetUtility.ResetAction.RESET_LAYERS,
-                    ResetUtility.ResetAction.RESET_CACHE, ResetUtility.ResetAction.RESET_OSM_DROID
+                    ResetUtility.ResetAction.RESET_PREFERENCES,
+                    ResetUtility.ResetAction.RESET_INSTANCES,
+                    ResetUtility.ResetAction.RESET_FORMS,
+                    ResetUtility.ResetAction.RESET_LAYERS,
+                    ResetUtility.ResetAction.RESET_CACHE,
+                    ResetUtility.ResetAction.RESET_OSM_DROID
             );
 
-            Context targetContext = InstrumentationRegistry.getInstrumentation().getTargetContext();
-            new ResetUtility().reset(targetContext, resetActions);
+            new StorageStateProvider().disableUsingScopedStorage();
+            new ResetUtility().reset(context, resetActions);
+            new StorageStateProvider().enableUsingScopedStorage();
+            new ResetUtility().reset(context, resetActions);
+
+            // Setup storage location for tests
+            if (useScopedStorage) {
+                new StorageStateProvider().enableUsingScopedStorage();
+            } else {
+                new StorageStateProvider().disableUsingScopedStorage();
+            }
+
+            // Any dependencies (PropertyManager for instance) will already have been
+            // passed to JavaRosa so make sure everything is reset
+            ((Collect) context.getApplicationContext()).initializeJavaRosa();
```

Comment:
I didn't realize that applications were never recreated in Espresso tests before @seadowg explained this one to me. So `Collect.onCreate` does not run for every test.